File: app/src/main/java/com/example/lucas/haushaltsmanager/Entities/Color.java
package com.example.lucas.haushaltsmanager.Entities;

import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.ColorInt;

import java.util.Random;
import java.util.regex.Pattern;

public class Color extends android.graphics.Color implements Parcelable {
    public static final Creator<Color> CREATOR = new Creator<Color>() {

        @Override
        public Color createFromParcel(Parcel in) {

            return new Color(in);
        }

        @Override
        public Color[] newArray(int size) {

            return new Color[size];
        }
    };
    static final String VALID_COLOR_PATTERN = "#([0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})";
    private String color;

    public Color(String color) {
        if (!assertIsColorString(color)) {
            throw new IllegalArgumentException(String.format("Could not create Color from: '%s'", color));
        }

        this.color = color;
    }

    public Color(@ColorInt int color) {
        this.color = "#" + Integer.toHexString(color);
    }

    private Color(Parcel source) {
        color = source.readString();
    }

    public static Color random() {
        Random random = new Random();

        int nextInt = random.nextInt(0xffffff + 1);

        return new Color(String.format("#%06x", nextInt));
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Color)) {
            return false;
        }

        Color other = (Color) o;

        return other.color.equals(color);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(color);
    }

    public String getColorString() {
        return color;
    }

    @ColorInt
    public int getColorInt() {
        return parseColor(color);
    }

    // TODO: Könnte getBrightness Funktion enthalten

    private boolean assertIsColorString(String color) {
        return Pattern.compile(VALID_COLOR_PATTERN)
                .matcher(color)
                .matches();
    }
}

File: app/src/main/java/com/example/lucas/haushaltsmanager/entities/Color.kt
package com.example.lucas.haushaltsmanager.entities

import android.graphics.Color
import android.os.Parcelable
import androidx.annotation.ColorInt
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.util.*
import java.util.regex.Pattern

@Parcelize
class Color(val color: String) : Parcelable, Color() {
    @IgnoredOnParcel
    val validColorPttrn = "^#[0-9a-fA-F]{8}\$|#[0-9a-fA-F]{6}\$|#[0-9a-fA-F]{4}\$|#[0-9a-fA-F]{3}\$"

    init {
        guardAgainstInvalidColorString(color)
    }

    constructor(@ColorInt color: Int) : this(
        "#${Integer.toHexString(color)}"
    )

    companion object {
        @JvmStatic
        fun black() = Color("#000000")

        @JvmStatic
        fun white() = Color("#FFFFFF")

        @JvmStatic
        fun random(): com.example.lucas.haushaltsmanager.entities.Color {
            val randomInt = Random().nextInt(0xffffff + 1)

            return Color(String.format("#%06x", randomInt))
        }
    }

    fun getColorInt() = parseColor(color)

    private fun guardAgainstInvalidColorString(colorString: String) {
        if (assertIsColorString(colorString)) {
            return
        }

        throw IllegalArgumentException("Could not create Color from: '%s'".format(colorString))
    }

    private fun assertIsColorString(color: String): Boolean {
        return Pattern.compile(validColorPttrn)
            .matcher(color)
            .matches()
    }

    override fun equals(other: Any?): Boolean {
        if (other !is com.example.lucas.haushaltsmanager.entities.Color) {
            return false;
        }

        return color == other.color;
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + color.hashCode()
        result = 31 * result + validColorPttrn.hashCode()
        return result
    }
}